### PR TITLE
Query: Implement inheritance support for relational layer

### PR DESCRIPTION
**Pull request by smitpatel**
- Add discriminator predicate in SelectExpression (filed #15808 to remove redundant)
- Add TPH support in materializer. This happens at core level now
- Add translation for is operator inside lambda
- Add translation for OfType method

### CODE REVIEW

**Comment by smitpatel** [comment]
@roji - This involves some breaking changes or disabled tests which has overlap with FromSql.

File: src/EFCore.Relational/Query/Pipeline/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 1]
```diff
@@ -511,7 +512,59 @@ protected override ShapedQueryExpression TranslateMin(ShapedQueryExpression sour
...
+
+                var derivedType = entityType.GetDerivedTypes().SingleOrDefault(et => et.ClrType == resultType);
+                if (derivedType != null)
+                {
+                    var selectExpression = (SelectExpression)source.QueryExpression;
+                    var concreteEntityTypes = derivedType.GetConcreteTypesInHierarchy().ToList();
+                    var discriminatorColumn = selectExpression
+                        .BindProperty(entityShaperExpression.ValueBufferExpression, entityType.GetDiscriminatorProperty());
+
+                    SqlExpression predicate = null;
```
Comment (maumar) [inline src/EFCore.Relational/Query/Pipeline/RelationalQueryableMethodTranslatingExpressionVisitor.cs]: nit: var predicate = ... ? ... : ...

File: src/EFCore.Relational/Query/Pipeline/SqlExpressionFactory.cs [hunk 2]
```diff
@@ -500,6 +503,68 @@ public SqlConstantExpression Constant(object value, RelationalTypeMapping typeMa
             return new SqlConstantExpression(Expression.Constant(value), typeMapping);
         }
 
+        public SelectExpression Select(SqlExpression projection)
+        {
+            var selectExpression = new SelectExpression(
+                null,
```
Comment (maumar) [inline src/EFCore.Relational/Query/Pipeline/SqlExpressionFactory.cs]: alias:

File: src/EFCore.Relational/Query/Pipeline/RelationalQueryableMethodTranslatingExpressionVisitor.cs (hunk 1 above)
Comment (roji) [reply src/EFCore.Relational/Query/Pipeline/RelationalQueryableMethodTranslatingExpressionVisitor.cs]: And maybe even move into ApplyPredicate() below

File: src/EFCore.Relational/Query/Pipeline/RelationalSqlTranslatingExpressionVisitor.cs [hunk 3]
```diff
@@ -79,44 +80,95 @@ protected override Expression VisitMember(MemberExpression memberExpression)
...
-            var selectExpression = (SelectExpression)entityShaper.ValueBufferExpression.QueryExpression;
+            return ((SelectExpression)entityShaper.ValueBufferExpression.QueryExpression)
+                .BindProperty(entityShaper.ValueBufferExpression, property);
+        }
+
+        protected override Expression VisitTypeBinary(TypeBinaryExpression typeBinaryExpression)
+        {
+            if (typeBinaryExpression.NodeType == ExpressionType.TypeIs)
+            {
+                if (typeBinaryExpression.Expression is EntityShaperExpression entityShaperExpression)
```
Comment (roji) [inline src/EFCore.Relational/Query/Pipeline/RelationalSqlTranslatingExpressionVisitor.cs]: nit: combine into condition above

File: src/EFCore.Relational/Query/Pipeline/RelationalSqlTranslatingExpressionVisitor.cs [hunk 4]
```diff
@@ -79,44 +80,95 @@ protected override Expression VisitMember(MemberExpression memberExpression)
...
+                        return _sqlExpressionFactory.Constant(true);
+                    }
+
+                    var derivedType = entityType.GetDerivedTypes().SingleOrDefault(et => et.ClrType == typeBinaryExpression.TypeOperand);
+                    if (derivedType != null)
+                    {
+                        var concreteEntityTypes = derivedType.GetConcreteTypesInHierarchy().ToList();
+                        var discriminatorColumn = BindProperty(entityShaperExpression, entityType.GetDiscriminatorProperty());
+
+                        if (concreteEntityTypes.Count == 1)
```
Comment (roji) [inline src/EFCore.Relational/Query/Pipeline/RelationalSqlTranslatingExpressionVisitor.cs]: nit: ternary operator

File: src/EFCore.Relational/Query/Pipeline/RelationalSqlTranslatingExpressionVisitor.cs [hunk 5]
```diff
@@ -79,44 +80,95 @@ protected override Expression VisitMember(MemberExpression memberExpression)
...
+                if (typeBinaryExpression.Expression is EntityShaperExpression entityShaperExpression)
+                {
+                    var entityType = entityShaperExpression.EntityType;
+                    if (entityType.GetAllBaseTypesInclusive().Any(et => et.ClrType == typeBinaryExpression.TypeOperand))
+                    {
+                        return _sqlExpressionFactory.Constant(true);
+                    }
+
+                    var derivedType = entityType.GetDerivedTypes().SingleOrDefault(et => et.ClrType == typeBinaryExpression.TypeOperand);
+                    if (derivedType != null)
```
Comment (roji) [inline src/EFCore.Relational/Query/Pipeline/RelationalSqlTranslatingExpressionVisitor.cs]: nit: inverse the logic, first checking `derivedType == null`, would reduce nesting afterwards.

**Comment by smitpatel** [comment]
@roji @maumar - Anything apart from nit? Approach looks fine?

**Comment by roji** [comment]
Will try to do a more in-depth review tomorrow, but the little I saw made total sense..

**Comment by smitpatel** [comment]
I will rebase on master then.

File: src/EFCore/Query/Pipeline/ProjectionBindingExpression.cs [hunk 6]
```diff
@@ -26,7 +28,7 @@ public ProjectionBindingExpression(Expression queryExpression, int index, Type t
 
         public Expression QueryExpression { get; }
         public ProjectionMember ProjectionMember { get; }
-        public int Index { get; }
+        public object Index { get; }
```
Comment (maumar) [inline src/EFCore/Query/Pipeline/ProjectionBindingExpression.cs]: split into Index (for scalar) and IndexMap (for entities)? its bit confusing as it is

File: src/EFCore.Relational/Query/Pipeline/RelationalSqlTranslatingExpressionVisitor.cs [hunk 7]
```diff
@@ -79,44 +80,95 @@ protected override Expression VisitMember(MemberExpression memberExpression)
...
+                }
+            }
 
-            return selectExpression.BindProperty(entityShaper.ValueBufferExpression, property);
+            return null;
         }
 
         protected override Expression VisitMethodCall(MethodCallExpression methodCallExpression)
         {
             if (methodCallExpression.Method.IsEFPropertyMethod())
```
Comment (roji) [inline src/EFCore.Relational/Query/Pipeline/RelationalSqlTranslatingExpressionVisitor.cs]: Can use the new TryGetEFPropertyArguments (just merged to master)

File: src/EFCore.Relational/Query/Pipeline/RelationalSqlTranslatingExpressionVisitor.cs (hunk 5 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/Pipeline/RelationalSqlTranslatingExpressionVisitor.cs]: After converting to ternary above there is no nesting.

**Pull request merged**